Mathlib/LinearAlgebra/Projection.lean

```diff
@@ -424,6 +425,41 @@ theorem eq_conj_prod_map' {f : E →ₗ[R] E} (h : IsProj p f) :
   · simp only [coe_prodEquivOfIsCompl, comp_apply, coe_inr, coprod_apply, map_zero,
       coe_subtype, zero_add, map_coe_ker, prodMap_apply, zero_apply, add_zero]
 
+theorem isIdempotentElem {R V : Type*} [Semiring R] [AddCommMonoid V] [Module R V]
+    {T : V →ₗ[R] V} {U : Submodule R V} (h : IsProj U T) : IsIdempotentElem T := by
+  ext; exact h.2 _ (h.1 _)
+
+protected theorem range
+    {R V : Type*} [Semiring R] [AddCommMonoid V] [Module R V] {T : V →ₗ[R] V} {U : Submodule R V}
+    (h : IsProj U T) : range T = U := by
+  ext x
+  refine ⟨fun ⟨y, hy⟩ => ?_, fun hx => ⟨x, h.map_id x hx⟩⟩
+  rw [← hy]
+  exact h.map_mem y
+
+protected theorem top (S M : Type*) [Semiring S] [AddCommMonoid M] [Module S M] :
+    IsProj (⊤ : Submodule S M) (id (R := S)) := ⟨fun _ ↦ trivial, fun _ ↦ congrFun rfl⟩
+
+theorem _root_.LinearMap.subtype_top_comp_codRestrict_top_eq_id
+    {S M : Type*} [Semiring S] [AddCommMonoid M] [Module S M] :
+    (Submodule.subtype ⊤).comp (IsProj.top S M).codRestrict = LinearMap.id := rfl
+
+theorem subtype_comp_codRestrict {S M : Type*} [Semiring S] [AddCommMonoid M]
+    [Module S M] {U : Submodule S M} {f : M →ₗ[S] M} (hf : IsProj U f) :
+    (Submodule.subtype U).comp hf.codRestrict = f := rfl
+
+theorem submodule_eq_top_iff {S M : Type*} [Semiring S] [AddCommMonoid M] [Module S M]
+    {f : M →ₗ[S] M} {U : Submodule S M} (hf : IsProj U f) :
+    U = (⊤ : Submodule S M) ↔ (Submodule.subtype _).comp hf.codRestrict = LinearMap.id := by
```

Comment:
This seems like a strange theorem to me, in the sense that the criterion that it gives for `U = ⊤` seems unlikely to be helpful in practice. Can you give some context? I would have expected something like this:
```suggestion
theorem submodule_eq_top_iff {S M : Type*} [Semiring S] [AddCommMonoid M] [Module S M]
    {f : M →ₗ[S] M} {U : Submodule S M} (hf : IsProj U f) :
    U = (⊤ : Submodule S M) ↔ f = id := by
  refine ⟨?_, fun h ↦ ?_⟩
  · rintro rfl
    ext x
    exact hf.map_id x mem_top
  · rw [← hf.range, h, range_id]
```